### ISSUE TITLE
CSS ::first-letter not supported (some drop caps not displaying)

### ISSUE THREAD

**Issue by mallomar** [koreader/koreader#9142]
* KOReader version: 2022.03.1-35
* Device: Onyx Boox Note Air 2

#### Issue:
Forgive me if this is a toggle or something but I can't figure it out and trying the [steps to resolve it in the only similar bug report I found didn't help. ](https://github.com/koreader/koreader/issues/8284) So, I bought a book, _Cribsheet_, from Amazon. On the Amazon web viewer it shows with drop caps:
<img width="1111" alt="Cribsheet_Kindle_web" src="https://user-images.githubusercontent.com/40035402/170853152-45e0c3b3-77a4-41e8-a81a-240396136293.png">

I brought the azw3 file of the book into Calibre and converted it to an epub. The epub in Calibre shows with drop caps:
<img width="639" alt="Cribsheet_Calibre" src="https://user-images.githubusercontent.com/40035402/170853225-bba9db0d-132d-4b52-87f4-5850d4efbad2.png">

But when I added the file from Calibre to my Note Air 2 and open it in KOReader, no drop caps show: 
![Screenshot_20220529-005125](https://user-images.githubusercontent.com/40035402/170853376-2a5da62b-8339-4ac0-b5a9-f3d94725aee9.png)

I tried looking through KOReader options in the various menus but I don't see anything that would toggle this and again the DPI setting doesn't appear to affect it. Any other ideas? Thanks!

#### Steps to reproduce:
Open file in KOReade
[Cribsheet_ A Data-Driven Guide to Better, - Emily Oster.epub.zip](https://github.com/koreader/koreader/files/8792700/Cribsheet_.A.Data-Driven.Guide.to.Better.-.Emily.Oster.epub.zip)
r

**Comment by mergen3107** [comment]
Have you tried enabling embedded styles? It’s in the bottom menu, next to embedded fonts

**Comment by Frenzie** [comment]
It's likely using ::first-letter, which is currently not supported.

**Comment by poire-z** [comment]
Indeed, not supported (and no obvious solution, the way crengine works, for it to be supported soon):

![image](https://user-images.githubusercontent.com/24273478/170857080-4d8cd5e6-f8a0-4ce0-8ed9-62d60e188da9.png)

![image](https://user-images.githubusercontent.com/24273478/170857055-1a14e0cc-b615-46a6-8f00-819ae41b7e65.png)

**Comment by poire-z** [comment]
Although `:first-letter` is probably easier to implement (but not straightforward without complications :/) than `:first-line` : 
On loading, wrap the first char of the text node in a `<firstLetterBox>` element (like we have `<pseudoElem>`).
For `:first-line`, we'd need to measure the text length to know where to break and wrap the text node (and if multiple inline nodes on the first line, this becomes a nightmare), which is done much later.
But even for `:first-letter`, it's complicated if there are inline nodes and the first letter is in some nested inline node. Moreover, then, any highlight/xpointer made later in that paragraph would have their index shifted by 1 (as the first letter is no longer in the same text node). Quite too complicated to give it a go, for something that is just cosmetics :/

**Comment by mallomar** [comment]
> Although `:first-letter` is probably easier to implement (but not straightforward without complications :/) than `:first-line` : On loading, wrap the first char of the text node in a `<firstLetterBox>` element (like we have `<pseudoElem>`). For `:first-line`, we'd need to measure the text length to know where to break and wrap the text node (and if multiple inline nodes on the first line, this becomes a nightmare), which is done much later. But even for `:first-letter`, it's complicated if there are inline nodes and the first letter is in some nested inline node. Moreover, then, any highlight/xpointer made later in that paragraph would have their index shifted by 1 (as the first letter is no longer in the same text node). Quite too complicated to give it a go, for something that is just cosmetics :/

Thanks, so you think it likely won't be supported anytime soon, if at all? I was surprised to see it working in Calibre but not in KOReader.

**Comment by poire-z** [comment]
> so you think it likely won't be supported anytime soon, if at all? 

Right.

>  I was surprised to see it working in Calibre but not in KOReader.

Thanks for your trust :) but Calibre must use some common full fledge web renderer like the Chrome or Webkit/Safari browser engines that large teams have been working on for decades - while KOreader use its own rendering engine of Russian origin, where every new little feature has to be added with my little fingers (and love) :)

**Comment by SilverGreen93** [comment]
I have the same issue in KOReader 2023.10.
The `:first-letter` CSS seems to be ignored.
I attach another example epub snippet which shows correctly in Sigil and Pocketbook internal reader app, but does not display correctly in KOReader.
[first-letter.zip](https://github.com/koreader/koreader/files/13724680/first-letter.zip)

**Comment by SilverGreen93** [comment]
Confirmed fixed in v2026.03
